### ISSUE TITLE
new TUI using a bit more CPU than old TUI

### ISSUE THREAD

**Issue by aktau** [neovim/neovim#2036]
This is like a tiny "test" I often do, which is as unscientific as it gets, but this is the first time I run it with the new TUI and I noticed "a regression" of sorts. Steps I use:

1. Open tmux
2. Open htop in a split screen
3. Do "nvim nvim/eval.c" in the other split
4. Watch nvim use 10-13% CPU

Doing the same steps with vanilla vim yields about 6-9% CPU. This used to be the same for neovim with the old terminal UI as well. Perhaps this is expected and I'm just adding noise, but just for reference. I'm at commit: e9c5f96964.

I'll try profiling with OSX instruments to see if I pick up on something suspicious, but it could just be a property of the new TUI codepaths, which I haven't studied sadly.

**Comment by tarruda** [comment]
Some redundancy was introduced with the new UI architecture: Two internal screens need to be updated for every redraw(one is maintained in nvim core and the other by the UI layer). This is required for the client-server architecture, but maybe later we can optimize this for the case where the UI and nvim core runs in the same process.

**Comment by aktau** [comment]
Just a small note to myself: yesterday I thought I had made a huge mistake because I was comparing self-built Neovim with homebrew-compiled Vim. The Neovim binary was a debugging binary. However, I recompiled with -O2 -flto and the result was exactly the same. Which seems to point to the double buffering causing it.

**Comment by aktau** [comment]
False alarm. I'm sorry. I had made 2 more errors while making the last statement:

1. I had recompiled without doing `rm -rf build`, causing only the linking step to be redone
2. I didn't run `make install`. Since I now always just run `nvim` to avoid runtime issues, I need to learn to do an install before actually running something

After doing it correctly, it appears that CPU usage between vim and nvim is very comparable indeed. Sorry for sounding the false alarm, and congrats on making a very clean UI/TUI interface. It's miles (leagues, lightyears) ahead of the old stuff.
